### PR TITLE
feat: add conversion of conv from onnx to tosa

### PR DESCRIPTION
**Pull request by maxbartel**
Signed-off-by: Maximilian Bartel <maximilian.bartel@amd.com>

### CODE REVIEW

File: src/Conversion/ONNXToTOSA/Math/Conv2D.cpp [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+
+using namespace mlir;
+
+namespace onnx_mlir {
+
+namespace {
+
+/// When setting the groups parameters, we have to create multiple conv2d ops
+/// where the input, kernel and bias is a slice of the original inputs.
+/// Afterwards we have to concat the results into a single tensor
```
Comment (ljfitz) [inline src/Conversion/ONNXToTOSA/Math/Conv2D.cpp:37]:
```suggestion
/// tosa.conv2d does not support dividing the channels into groups.
/// When the onnx operator requires this, we have to create multiple ops
/// where the input, kernel and bias is a slice of the original inputs.
/// Afterwards we have to concat the results of these into a single tensor.
```

File: src/Conversion/ONNXToTOSA/Math/Conv2D.cpp [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+Value createConvInGroups(PatternRewriter &rewriter, Operation *op,
+    TosaBuilder &tosaBuilder, Type &resultType,
+    const llvm::ArrayRef<int64_t> weightShape, Value &newInput,
+    Value &newWeight, Value &bias, const int64_t groups,
+    DenseI64ArrayAttr &pads, DenseI64ArrayAttr &strides,
+    DenseI64ArrayAttr &dilations) {
+  // Set up constants outside of loop
+  const int64_t sizeOfSliceInput = weightShape[1];
+  const int64_t sizeOfSliceKernel = weightShape[0] / groups;
```
Comment (ljfitz) [inline src/Conversion/ONNXToTOSA/Math/Conv2D.cpp:46]: I'm pretty sure the definition requires that `weightShape[0] % groups == 0` but wonder if you want to assert is with a comment that legally this must hold? I'm not certain if it is needed - I personally just get paranoid reading code with integer `/` without the `%` being used/checked.

File: src/Conversion/ONNXToTOSA/Math/Conv2D.cpp [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    IndexExprBuilderForTosa createTosaIE(rewriter, convOp->getLoc());
+    ONNXConvOpShapeHelper shapeHelper(op, operands, &createTosaIE);
+    shapeHelper.computeShapeAndAssertOnFailure();
+
+    auto weightShape = weightType.getShape();
+
+    Type resultType = convOp.getResult().getType();
+
+    if (inputType.getShape().size() != 4) {
+      return rewriter.notifyMatchFailure(convOp, "TOSA only supports conv 2d");
```
Comment (ljfitz) [inline src/Conversion/ONNXToTOSA/Math/Conv2D.cpp:118]:
```suggestion
      return rewriter.notifyMatchFailure(convOp, "only 2d tensor support is implemented");
```

**Review by ljfitz — APPROVED** [review]

File: src/Conversion/ONNXToTOSA/Math/Conv2D.cpp (hunk 2 above)
Comment (maxbartel) [reply src/Conversion/ONNXToTOSA/Math/Conv2D.cpp:46]: The ONNX verifier actually checks for that and will not even read the IR if this is incorrect

**Pull request closed without merging**